#### Introduction

Welcome to Priceless Wheels! In this project, our goal is to build a model that can accurately predict the price of a used vehicle based on various factors such as make, model, year, mileage, and condition. The automobile industry is one of the largest and most competitive industries in the world, with millions of vehicles being sold each year. The price of a vehicle can have a significant impact on a consumer's purchasing decision and it is important for both buyers and sellers to have an understanding of the market value of a vehicle. By using machine learning algorithms and data analysis, we aim to provide a reliable and robust model that can assist in determining the fair market value of a vehicle. Join us on this exciting journey as we delve into the world of vehicle price prediction.

#### About the data

This data is scrapped from https://www.cardekho.com/. This data is meant for research and academic purposes only and is **not meant for commercial use**. This dataset contains about 38000 (thirty eight thousand) used cars listed in CarDekho in India. Download the data from https://github.com/chats-bug/Priceless-Wheels/blob/data-cleaning/data/raw/cardekho_cars_2023_03_19_16_44_14.csv.

---

##### Importing the data and libraries

Let's start by importing the necessary libraries and the data.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O
import matplotlib.pyplot as plt
import seaborn as sns
sns.set() # setting seaborn default for plots

# removing scientific notation
pd.set_option('display.float_format', lambda x: '%.3f' % x)

Next, we will import the data and preview the first 5 rows

- Note: The ```usedCarSkuId``` is a unique identifier for each car. We would be using this column as the index of the dataframe
- Note: Since we are using github lfs for storing the CSV files, the link mentioned in the notebook will expire in some time. Please visit this [link](https://github.com/chats-bug/Priceless-Wheels/blob/data-cleaning/data/raw/cardekho_cars_2023_03_19_16_44_14.csv) to get the data

In [2]:
# Read the data from the following github link: https://media.githubusercontent.com/media/chats-bug/Priceless-Wheels/data-cleaning/data/raw/cardekho_cars_2023_03_19_16_44_14.csv?token=ARCDTMXTKLJI6OWOM7UVQBLEC5UKY

file_link = "https://media.githubusercontent.com/media/chats-bug/Priceless-Wheels/data-cleaning/data/raw/cardekho_cars_2023_03_19_16_44_14.csv?token=ARCDTMQJITVJHMJYPHJZ64DEC6GRI"
df = pd.read_csv(file_link, index_col="usedCarSkuId")

# sanity check
df.head()

/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3553: DtypeWarning: Columns (25,43) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


position            loc  myear  \
usedCarSkuId                                                           
7111bf25-97af-47f9-867b-40879190d800         1    Gomti Nagar   2016   
c309efc1-efaf-4f82-81ad-dcb38eb36665         2  Borivali West   2015   
7609f710-0c97-4f00-9a47-9b9284b62d3a         3         JASOLA   2015   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d         4         jasola   2013   
b1eab99b-a606-48dd-a75b-57feb8a9ad92         5  mumbai g.p.o.   2022   

                                             bt      tt   ft        km  ip  \
usedCarSkuId                                                                 
7111bf25-97af-47f9-867b-40879190d800  Hatchback  Manual  CNG    69,162   0   
c309efc1-efaf-4f82-81ad-dcb38eb36665  Hatchback  Manual  CNG    45,864   0   
7609f710-0c97-4f00-9a47-9b9284b62d3a      Sedan  Manual  CNG    81,506   0   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d  Hatchback  Manual  CNG  1,15,893   0   
b1eab99b-a606-48dd-a75b-57feb8a9ad92        MUV  Manual  CNG    18,900   0   

                                                                                     pi  \
usedCarSkuId                                                                              
7111bf25-97af-47f9-867b-40879190d800  https://images10.gaadi.com/usedcar_image/origi...   
c309efc1-efaf-4f82-81ad-dcb38eb36665  https://images10.gaadi.com/usedcar_image/origi...   
7609f710-0c97-4f00-9a47-9b9284b62d3a  https://images10.gaadi.com/usedcar_image/origi...   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d                                                NaN   
b1eab99b-a606-48dd-a75b-57feb8a9ad92  https://images10.gaadi.com/usedcar_image/origi...   

                                                                                 images  \
usedCarSkuId                                                                              
7111bf25-97af-47f9-867b-40879190d800  [{'img': 'https://images10.gaadi.com/usedcar_i...   
c309efc1-efaf-4f82-81ad-dcb38eb36665  [{'img': 'https://images10.gaadi.com/usedcar_i...   
7609f710-0c97-4f00-9a47-9b9284b62d3a  [{'img': 'https://images10.gaadi.com/usedcar_i...   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d                                      [{'img': ''}]   
b1eab99b-a606-48dd-a75b-57feb8a9ad92  [{'img': 'https://images10.gaadi.com/usedcar_i...   

                                      ...  owner_type  price_segment_new  \
usedCarSkuId                          ...                                  
7111bf25-97af-47f9-867b-40879190d800  ...       first        2lakh-5lakh   
c309efc1-efaf-4f82-81ad-dcb38eb36665  ...       first        2lakh-5lakh   
7609f710-0c97-4f00-9a47-9b9284b62d3a  ...      second        2lakh-5lakh   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d  ...      second        2lakh-5lakh   
b1eab99b-a606-48dd-a75b-57feb8a9ad92  ...       first            10+lakh   

                                                   template_name_new  \
usedCarSkuId                                                           
7111bf25-97af-47f9-867b-40879190d800  used cardetail v2/corporate/13   
c309efc1-efaf-4f82-81ad-dcb38eb36665  used cardetail v2/corporate/13   
7609f710-0c97-4f00-9a47-9b9284b62d3a  used cardetail v2/corporate/13   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d  used cardetail v2/corporate/13   
b1eab99b-a606-48dd-a75b-57feb8a9ad92               used cardetail v2   

                                               page_template  \
usedCarSkuId                                                   
7111bf25-97af-47f9-867b-40879190d800  Used Car > Detail Page   
c309efc1-efaf-4f82-81ad-dcb38eb36665  Used Car > Detail Page   
7609f710-0c97-4f00-9a47-9b9284b62d3a  Used Car > Detail Page   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d  Used Car > Detail Page   
b1eab99b-a606-48dd-a75b-57feb8a9ad92  Used Car > Detail Page   

                                     template_Type_new  experiment  \
usedCarSkuId                                                         
7111bf25-97af-47f9-867b-40879190d800              used     control   
c309efc1-efaf

The data has imported with some mixed data types. We will need to convert the data types as we explore the data.

In [3]:
# checking the shape of the dataset
shape = df.shape

print(f"The shape of the dataframe is {shape[0]} rows and {shape[1]} columns")

The shape of the dataframe is 37814 rows and 139 columns


As we can see, there are 37,814 rows and 139 columns in the dataset.

Let's look at the data types of the columns.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 37814 entries, 7111bf25-97af-47f9-867b-40879190d800 to a96fbcd7-c183-4829-ae97-b2581afe4bac
Columns: 139 entries, position to Ground Clearance Unladen
dtypes: bool(3), float64(10), int64(28), object(98)
memory usage: 39.6+ MB


Now comes the difficult task. Go through the feature dictionary and remove the unwanted columns - features which might be repeated but not totally apparent. Note down the columns which makes sense to keep and are not repeating and delete the rest.

In [5]:
columns_to_keep = [
    "loc",
    "myear",
    "bt",
    "tt",
    "ft"
    "km",
    "ip",
    "images",
    "imgCount",
    "threesixty",
    "dvn",
    "oem",
    "model",
    "variantName",
    "city_x",
    "pu",
    "discountValue",
    "utype",
    "carType", 
    "top_features",
    "comfort_features",
    "interior_features",
    "exterior_features",
    "safety_features",
    "Color",
    "Engine Type",
    "Max Power",
    "Max Torque",
    "No of Cylinder",
    "Values per Cylinder",
    "Value Configuration",
    "BoreX Stroke",
    "Turbo Charger",
    "Super Charger",
    "Length",
    "Width",
    "Height",
    "Wheel Base",
    "Front Tread",
    "Rear Tread",
    "Kerb Weight",
    "Gross Weight",
    "Gear Box",
    "Drive Type",
    "Seating Capacity",
    "Steering Type",
    "Turning Radius",
    "Front Brake Type",
    "Rear Brake Type",
    "Top Speed",
    "Acceleration",
    "Tyre Type",
    "No Door Numbers",
    "Cargo Volumn",
    "model_type_new",
    "state",
    "owner_type",
    "exterior_color",
    "Fuel Suppy System",
    "Compression Ratio",
    "Alloy Wheel Size",
    "Ground Clearance Unladen",
]

df.drop([x for x in df.columns if x not in columns_to_keep], axis=1, inplace=True)
print(f"After dropping some unnecessary columns, the dataset now has {df.shape[1]} columns. These columns are hand picked and will be be further analyzed.")

After dropping some unnecessary columns, the dataset now has 60 columns. These columns are hand picked and will be be further analyzed.


---

### Data cleaning and wrangling

* Dropping duplicate rows
* Checking for multicollinearity and determining how to handle it
* Filling in missing values/NaN values
* Fixing the values and data types of the columns
* Dropping irrelevant columns for the model
* Saving the cleaned data for the next part of the project

---

### Duplicates

We know that there should be no duplicates in the data. They are checked for and removed at the time of data collection. *(Trust me, I collected it myself 😉 )*

In [6]:
# checking for duplicate rows
duplucate_rows = df.duplicated().sum()

print(f"The number of duplicate rows are {duplucate_rows}.")

The number of duplicate rows are 1.


Somehow, one duplicate row sneaked in (that's awkward 🫣)! Let's check the rows.

In [7]:
# setting the duplicated index
duplicate_index = df.duplicated(keep=False)

# calling the duplicated index in a dataframe
df.loc[duplicate_index, :].sort_index()

loc  myear         bt      tt  ip  \
usedCarSkuId                                                              
aa39e640-6183-4379-a517-9f5b2458b2a5  NaN   2014  Hatchback  Manual   0   
ce8b30d8-c438-4a2f-bce1-5c1887a95495  NaN   2014  Hatchback  Manual   0   

                                             images  imgCount  threesixty  \
usedCarSkuId                                                                
aa39e640-6183-4379-a517-9f5b2458b2a5  [{'img': ''}]         0       False   
ce8b30d8-c438-4a2f-bce1-5c1887a95495  [{'img': ''}]         0       False   

                                                        dvn     oem  ...  \
usedCarSkuId                                                         ...   
aa39e640-6183-4379-a517-9f5b2458b2a5  Maruti Swift VDI BSIV  Maruti  ...   
ce8b30d8-c438-4a2f-bce1-5c1887a95495  Maruti Swift VDI BSIV  Maruti  ...   

                                     No Door Numbers Cargo Volumn  \
usedCarSkuId                                                        
aa39e640-6183-4379-a517-9f5b2458b2a5           5.000   204-liters   
ce8b30d8-c438-4a2f-bce1-5c1887a95495           5.000   204-liters   

                                     model_type_new          state  \
usedCarSkuId                                                         
aa39e640-6183-4379-a517-9f5b2458b2a5           used  uttar pradesh   
ce8b30d8-c438-4a2f-bce1-5c1887a95495           used  uttar pradesh   

                                      exterior_color owner_type  \
usedCarSkuId                                                      
aa39e640-6183-4379-a517-9f5b2458b2a5          Silver     second   
ce8b30d8-c438-4a2f-bce1-5c1887a95495          Silver     second   

                                     Fuel Suppy System Compression Ratio  \
usedCarSkuId                                                               
aa39e640-6183-4379-a517-9f5b2458b2a5              CRDI            17.6:1   
ce8b30d8-c438-4a2f-bce1-5c1887a95495              CRDI            17.6:1   

                                     Alloy Wheel Size Ground Clearance Unladen  
usedCarSkuId                                                                    
aa39e640-6183-4379-a517-9f5b2458b2a5              NaN                      NaN  
ce8b30d8-c438-4a2f-bce1-5c1887a95495              NaN                      NaN  

[2 rows x 60 columns]

As we can see, the rows are exactly the same. This means we can safely drop the row. Since the we indexed by the usedCarSkuId, we need to drop one of them by their index.

In [8]:
# making a copy with the duplicated rows dropped
df2 = df.drop_duplicates().copy()

# checking for duplicate rows in the new dataframe
dup = df2.duplicated().sum()

print(f"The number of duplicate rows are {dup}.")

The number of duplicate rows are 0.


In [9]:
# sanity check
df2.head()

loc  myear         bt      tt  \
usedCarSkuId                                                                    
7111bf25-97af-47f9-867b-40879190d800    Gomti Nagar   2016  Hatchback  Manual   
c309efc1-efaf-4f82-81ad-dcb38eb36665  Borivali West   2015  Hatchback  Manual   
7609f710-0c97-4f00-9a47-9b9284b62d3a         JASOLA   2015      Sedan  Manual   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d         jasola   2013  Hatchback  Manual   
b1eab99b-a606-48dd-a75b-57feb8a9ad92  mumbai g.p.o.   2022        MUV  Manual   

                                      ip  \
usedCarSkuId                               
7111bf25-97af-47f9-867b-40879190d800   0   
c309efc1-efaf-4f82-81ad-dcb38eb36665   0   
7609f710-0c97-4f00-9a47-9b9284b62d3a   0   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d   0   
b1eab99b-a606-48dd-a75b-57feb8a9ad92   0   

                                                                                 images  \
usedCarSkuId                                                                              
7111bf25-97af-47f9-867b-40879190d800  [{'img': 'https://images10.gaadi.com/usedcar_i...   
c309efc1-efaf-4f82-81ad-dcb38eb36665  [{'img': 'https://images10.gaadi.com/usedcar_i...   
7609f710-0c97-4f00-9a47-9b9284b62d3a  [{'img': 'https://images10.gaadi.com/usedcar_i...   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d                                      [{'img': ''}]   
b1eab99b-a606-48dd-a75b-57feb8a9ad92  [{'img': 'https://images10.gaadi.com/usedcar_i...   

                                      imgCount  threesixty  \
usedCarSkuId                                                 
7111bf25-97af-47f9-867b-40879190d800        15       False   
c309efc1-efaf-4f82-81ad-dcb38eb36665        15       False   
7609f710-0c97-4f00-9a47-9b9284b62d3a        15       False   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d         0       False   
b1eab99b-a606-48dd-a75b-57feb8a9ad92         6       False   

                                                            dvn     oem  ...  \
usedCarSkuId                                                             ...   
7111bf25-97af-47f9-867b-40879190d800     Maruti Wagon R LXI CNG  Maruti  ...   
c309efc1-efaf-4f82-81ad-dcb38eb36665   Maruti Celerio Green VXI  Maruti  ...   
7609f710-0c97-4f00-9a47-9b9284b62d3a  Honda Amaze S Plus i-VTEC   Honda  ...   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d     Maruti Wagon R LXI CNG  Maruti  ...   
b1eab99b-a606-48dd-a75b-57feb8a9ad92      Maruti Ertiga VXI CNG  Maruti  ...   

                                     No Door Numbers Cargo Volumn  \
usedCarSkuId                                                        
7111bf25-97af-47f9-867b-40879190d800           5.000   180-liters   
c309efc1-efaf-4f82-81ad-dcb38eb36665           5.000   235-litres   
7609f710-0c97-4f00-9a47-9b9284b62d3a           4.000   400-litres   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d           4.000          NaN   
b1eab99b-a606-48dd-a75b-57feb8a9ad92           5.000          NaN   

                                     model_type_new          state  \
usedCarSkuId                                                         
7111bf25-97af-47f9-867b-40879190d800           used  uttar pradesh   
c309efc1-efaf-4f82-81ad-dcb38eb36665           used    maharashtra   
7609f710-0c97-4f00-9a47-9b9284b62d3a           used          delhi   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d           used          delhi   
b1eab99b-a606-48dd-a75b-57feb8a9ad92           used    maharashtra   

                                      exterior_color owner_type  \
usedCarSkuId                                                      
7111bf25-97af-47f9-867b-40879190d800          Silver      first   
c309efc1-efaf-4f82-81ad-dcb38eb36665            Grey      first   
7609f710-0c97-4f00-9a47-9b9284b62d3a          Silver     second   
278b76e3-5539-4a5e-ae3e-353a2e3b6d7d          Silver     second   
b1eab99b-a606-48dd-a75b-57feb8a9ad92           White      first   

                                                   Fuel Suppy System  \
usedCarSkuId        

Great! Now that we have removed the duplicate rows, let's look at missing values

---

### Missing Null Values

In [10]:
df2.isnull().sum()

loc                          5850
myear                           0
bt                             19
tt                              0
ip                              0
images                          0
imgCount                        0
threesixty                      0
dvn                             0
oem                             0
model                           0
variantName                     0
city_x                          0
pu                              0
discountValue                   0
utype                           0
carType                         0
top_features                    0
comfort_features                0
interior_features               0
exterior_features               0
safety_features                 0
Color                         160
Engine Type                   868
Max Power                     231
Max Torque                    231
No of Cylinder                143
Values per Cylinder           228
Value Configuration          7831
BoreX Stroke  

There are several columns that have an exceptionally large amount of null values. This means that these columns will not be useful for our analysis because of the sheer amount of missing values.

Let's set a threshold of 50% and drop any columns above that threshold. This will allow us to remove the columns that have a large amount of missing values that can skew our analysis.

In [11]:
# Delete columns containing more than 50% null values
percentage = 50
min_count =  int(((100-percentage)/100)*df2.shape[0] + 1)
cars_df = df2.dropna(axis=1, thresh=min_count)

print(f"The columns with {percentage}% or more null values are: {df2.columns[df2.isna().sum()/df2.shape[0]*100 > percentage]}")
print(f"The new shape of the dataframe is {cars_df.shape[0]} rows and {cars_df.shape[1]} columns.")

The columns with 50% or more null values are: Index(['BoreX Stroke', 'Gross Weight', 'Compression Ratio',
       'Ground Clearance Unladen'],
      dtype='object')
The new shape of the dataframe is 37813 rows and 56 columns.


The results show that the following columns have been dropped:

* ```BoreX Stroke```
* ```Gross Weight```
* ```Compression Ratio```
* ```Ground Clearance Unladen```


Lets filter only the columns with null values.

In [12]:
# checking how many null values are in the dataframe that are over 0%
percentage = cars_df.isna().sum()/cars_df.shape[0]*100

pd.DataFrame(percentage[percentage > 0]).sort_values(by=0, ascending=False)

0
Rear Tread          41.983
Front Tread         41.980
Top Speed           39.764
Acceleration        36.993
Alloy Wheel Size    34.763
Kerb Weight         34.747
Value Configuration 20.710
Cargo Volumn        17.803
loc                 15.471
Turning Radius      15.466
Fuel Suppy System   14.551
Super Charger       14.143
Drive Type          11.890
Turbo Charger        5.755
Engine Type          2.296
Steering Type        2.137
Wheel Base           1.468
Gear Box             1.076
Width                0.981
Height               0.960
Length               0.957
Front Brake Type     0.865
Rear Brake Type      0.862
Tyre Type            0.680
Max Power            0.611
Max Torque           0.611
Values per Cylinder  0.603
Color                0.423
No of Cylinder       0.378
exterior_color       0.325
No Door Numbers      0.116
bt                   0.050
Seating Capacity     0.048

Now we can focus on each column individually to determine how to handle the missing values. There are so many! 😩

Let's make a copy of the dataframe so that we always have the original to go back to

In [13]:
cars_df2 = cars_df.copy()

---



#### Rear and Front Tread

Tread is the rubber on the tire that touches the road. As you drive, the tread will wear down. Since this indicates the same thing as km driven or how old the car is, it should not be very important for our ML model which we would be building.

Let's see how many null values we have in this column.

In [14]:
rt_null_percentage = cars_df2['Rear Tread'].isna().sum()/cars_df2.shape[0]*100
ft_null_percentage = cars_df2['Front Tread'].isna().sum()/cars_df2.shape[0]*100

print(f"There is about {rt_null_percentage} values missing in the Rear Tread column")
print(f"There is about {ft_null_percentage} values missing in the Front Tread column")

There is about 41.982915928384415 values missing in the Rear Tread column
There is about 41.98027133525507 values missing in the Front Tread column


We can see that both of them are missing ~40% of the values. Since we concluded they are not going to be super important for our models, we can drop them.

In [15]:
cars_df2.drop(['Rear Tread', 'Front Tread'], axis=1, inplace=True)
cars_df2.shape

(37813, 54)

We are now left with 54 columns. Let's look at the next feature.

---